File: apps/api/src/services/ai.service.ts
import { GoogleGenerativeAI } from '@google/generative-ai';
import { ContentNiche, ContentTone } from '@contentcraft/shared';

class AIService {
  private genAI: GoogleGenerativeAI;

  constructor() {
    if (!process.env.GEMINI_API_KEY) {
      throw new Error('GEMINI_API_KEY is required');
    }
    this.genAI = new GoogleGenerativeAI(process.env.GEMINI_API_KEY);
  }

  async generateCaption(params: {
    videoDescription: string;
    tone: ContentTone;
    niche: ContentNiche;
    includeHashtags?: boolean;
    maxLength?: number;
  }): Promise<{ caption: string; hashtags: string[] }> {
    const { videoDescription, tone, niche, includeHashtags = true, maxLength = 300 } = params;

    const model = this.genAI.getGenerativeModel({ model: 'gemini-pro' });

    const prompt = this.buildCaptionPrompt(videoDescription, tone, niche, includeHashtags, maxLength);

    try {
      const result = await model.generateContent(prompt);
      const response = await result.response;
      const text = response.text();

      return this.parseCaptionResponse(text);
    } catch (error) {
      console.error('Error generating caption:', error);
      throw new Error('Failed to generate caption');
    }
  }

  async generateHashtags(params: {
    content: string;
    niche: ContentNiche;
    count?: number;
  }): Promise<string[]> {
    const { content, niche, count = 10 } = params;

    const model = this.genAI.getGenerativeModel({ model: 'gemini-pro' });

    const prompt = `
Generate ${count} relevant hashtags for the following ${niche} content:

"${content}"

Requirements:
- Focus on ${niche} niche
- Mix of popular and niche-specific hashtags
- Include trending hashtags when relevant
- No spaces in hashtags
- Return as comma-separated list

Hashtags:`;

    try {
      const result = await model.generateContent(prompt);
      const response = await result.response;
      const text = response.text();

      return this.parseHashtagsResponse(text);
    } catch (error) {
      console.error('Error generating hashtags:', error);
      throw new Error('Failed to generate hashtags');
    }
  }

  async analyzeVideoContent(videoDescription: string): Promise<{
    niche: ContentNiche;
    tone: ContentTone;
    keywords: string[];
    summary: string;
  }> {
    const model = this.genAI.getGenerativeModel({ model: 'gemini-pro' });

    const prompt = `
Analyze the following video content and provide insights:

"${videoDescription}"

Please analyze and return in this exact format:
NICHE: [one of: FITNESS, FOOD, EDUCATION, LIFESTYLE, BUSINESS, TECH]
TONE: [one of: CASUAL, PROFESSIONAL, FUN, MOTIVATIONAL, EDUCATIONAL, TRENDY]
KEYWORDS: [comma-separated list of 5-10 relevant keywords]
SUMMARY: [2-3 sentence summary of the content]

Analysis:`;

    try {
      const result = await model.generateContent(prompt);
      const response = await result.response;
      const text = response.text();

      return this.parseAnalysisResponse(text);
    } catch (error) {
      console.error('Error analyzing video content:', error);
      throw new Error('Failed to analyze video content');
    }
  }

  async suggestOptimalPostingTimes(params: {
    niche: ContentNiche;
    timezone: string;
    platform: string;
  }): Promise<{ dayOfWeek: number; hour: number; reason: string }[]> {
    const { niche, timezone, platform } = params;

    const model = this.genAI.getGenerativeModel({ model: 'gemini-pro' });

    const prompt = `
Suggest the best posting times for ${niche} content on ${platform} for timezone ${timezone}.

Consider:
- Target audience activity patterns for ${niche}
- Platform-specific engagement patterns
- General social media best practices

Provide 3-5 optimal posting times with explanations.

Format:
DAY: [0-6, where 0=Sunday]
HOUR: [0-23]
REASON: [explanation]

Suggestions:`;

    try {
      const result = await model.generateContent(prompt);
      const response = await result.response;
      const text = response.text();

      return this.parsePostingTimesResponse(text);
    } catch (error) {
      console.error('Error suggesting posting times:', error);
      throw new Error('Failed to suggest posting times');
    }
  }

  private buildCaptionPrompt(
    description: string,
    tone: ContentTone,
    niche: ContentNiche,
    includeHashtags: boolean,
    maxLength: number
  ): string {
    const toneInstructions = {
      [ContentTone.CASUAL]: 'Use casual, conversational language. Be relatable and friendly.',
      [ContentTone.PROFESSIONAL]: 'Use professional, polished language. Be authoritative but approachable.',
      [ContentTone.FUN]: 'Use playful, energetic language. Include emojis and be entertaining.',
      [ContentTone.MOTIVATIONAL]: 'Use inspiring, uplifting language. Encourage and motivate the audience.',
      [ContentTone.EDUCATIONAL]: 'Use clear, informative language. Focus on teaching and explaining.',
      [ContentTone.TRENDY]: 'Use current slang and trending phrases. Be hip and relevant.',
    };

    const nicheInstructions = {
      [ContentNiche.FITNESS]: 'Focus on health, wellness, workout tips, and motivation.',
      [ContentNiche.FOOD]: 'Focus on recipes, cooking tips, ingredients, and food culture.',
      [ContentNiche.EDUCATION]: 'Focus on learning, teaching, knowledge sharing, and skill development.',
      [ContentNiche.LIFESTYLE]: 'Focus on daily life, personal experiences, and lifestyle choices.',
      [ContentNiche.BUSINESS]: 'Focus on entrepreneurship, productivity, and business insights.',
      [ContentNiche.TECH]: 'Focus on technology, innovation, and digital trends.',
    };

    return `
Create an engaging social media caption for this ${niche} content:

"${description}"

Requirements:
- Tone: ${tone} (${toneInstructions[tone]})
- Niche: ${niche} (${nicheInstructions[niche]})
- Maximum length: ${maxLength} characters
- ${includeHashtags ? 'Include relevant hashtags at the end' : 'No hashtags needed'}
- Include emojis where appropriate
- Make it engaging and likely to get likes/comments

${includeHashtags ? 'Format: Caption text\n\nHASHTAGS: #hashtag1 #hashtag2 #hashtag3' : 'Just return the caption text.'}

Caption:`;
  }

  private parseCaptionResponse(text: string): { caption: string; hashtags: string[] } {
    const lines = text.trim().split('\n');
    
    let caption = '';
    let hashtags: string[] = [];

    let isHashtagSection = false;
    
    for (const line of lines) {
      if (line.toUpperCase().includes('HASHTAGS:')) {
        isHashtagSection = true;
        const hashtagLine = line.replace(/HASHTAGS:/i, '').trim();
        if (hashtagLine) {
          hashtags = this.extractHashtags(hashtagLine);
        }
      } else if (isHashtagSection) {
        hashtags = [...hashtags, ...this.extractHashtags(line)];
      } else if (line.trim()) {
        caption += (caption ? ' ' : '') + line.trim();
      }
    }

    // If no hashtags section found, extract from caption
    if (hashtags.length === 0) {
      const captionHashtags = this.extractHashtags(caption);
      if (captionHashtags.length > 0) {
        // Remove hashtags from caption
        caption = caption.replace(/#\w+/g, '').trim();
        hashtags = captionHashtags;
      }
    }

    return {
      caption: caption.trim(),
      hashtags: hashtags.slice(0, 15), // Limit to 15 hashtags
    };
  }

  private parseHashtagsResponse(text: string): string[] {
    return this.extractHashtags(text);
  }

  private extractHashtags(text: string): string[] {
    const hashtags = text.match(/#\w+/g) || [];
    const cleanHashtags = text
      .split(/[,\s\n]+/)
      .map(tag => tag.trim())
      .filter(tag => tag.startsWith('#') || /^[a-zA-Z]/.test(tag))
      .map(tag => tag.startsWith('#') ? tag : `#${tag}`)
      .filter(tag => tag.length > 1);
    
    return [...new Set([...hashtags, ...cleanHashtags])];
  }

  private parseAnalysisResponse(text: string): {
    niche: ContentNiche;
    tone: ContentTone;
    keywords: string[];
    summary: string;
  } {
    const lines = text.split('\n');
    let niche = ContentNiche.LIFESTYLE;
    let tone = ContentTone.CASUAL;
    let keywords: string[] = [];
    let summary = '';

    for (const line of lines) {
      const upperLine = line.toUpperCase();
      
      if (upperLine.includes('NICHE:')) {
        const nicheValue = line.replace(/NICHE:/i, '').trim().toUpperCase();
        if (Object.values(ContentNiche).includes(nicheValue as ContentNiche)) {
          niche = nicheValue as ContentNiche;
        }
      } else if (upperLine.includes('TONE:')) {
        const toneValue = line.replace(/TONE:/i, '').trim().toUpperCase();
        if (Object.values(ContentTone).includes(toneValue as ContentTone)) {
          tone = toneValue as ContentTone;
        }
      } else if (upperLine.includes('KEYWORDS:')) {
        const keywordsText = line.replace(/KEYWORDS:/i, '').trim();
        keywords = keywordsText.split(',').map(k => k.trim()).filter(k => k.length > 0);
      } else if (upperLine.includes('SUMMARY:')) {
        summary = line.replace(/SUMMARY:/i, '').trim();
      }
    }

    return { niche, tone, keywords, summary };
  }

  private parsePostingTimesResponse(text: string): { dayOfWeek: number; hour: number; reason: string }[] {
    const lines = text.split('\n');
    const postingTimes: { dayOfWeek: number; hour: number; reason: string }[] = [];

    let currentDay: number | null = null;
    let currentHour: number | null = null;
    let currentReason = '';

    for (const line of lines) {
      const upperLine = line.toUpperCase();
      
      if (upperLine.includes('DAY:')) {
        if (currentDay !== null && currentHour !== null) {
          postingTimes.push({
            dayOfWeek: currentDay,
            hour: currentHour,
            reason: currentReason.trim(),
          });
        }
        const dayMatch = line.match(/DAY:\s*(\d+)/i);
        currentDay = dayMatch ? parseInt(dayMatch[1]) : null;
        currentReason = '';
      } else if (upperLine.includes('HOUR:')) {
        const hourMatch = line.match(/HOUR:\s*(\d+)/i);
        currentHour = hourMatch ? parseInt(hourMatch[1]) : null;
      } else if (upperLine.includes('REASON:')) {
        currentReason = line.replace(/REASON:/i, '').trim();
      }
    }

    if (currentDay !== null && currentHour !== null) {
      postingTimes.push({
        dayOfWeek: currentDay,
        hour: currentHour,
        reason: currentReason.trim(),
      });
    }

    return postingTimes.filter(pt => 
      pt.dayOfWeek >= 0 && pt.dayOfWeek <= 6 && 
      pt.hour >= 0 && pt.hour <= 23
    );
  }
}

export const aiService = new AIService(); 

File: backend/app/services/ai_service.py
"""
AI service for caption generation using Google Gemini API.
"""

import google.generativeai as genai
from typing import List, Dict, Any
import logging
import json
import re

from app.core.config import settings
from app.schemas.content import CaptionGenerationRequest, CaptionGenerationResponse
from app.models.content import Platform, ContentTone, ContentNiche

logger = logging.getLogger(__name__)

# Configure Gemini API
if settings.GEMINI_API_KEY:
    genai.configure(api_key=settings.GEMINI_API_KEY)


class AIService:
    """Service for AI-powered content generation."""
    
    def __init__(self):
        self.model = None
        if settings.GEMINI_API_KEY:
            try:
                self.model = genai.GenerativeModel('gemini-pro')
                logger.info("✅ Gemini AI model initialized successfully")
            except Exception as e:
                logger.error(f"❌ Failed to initialize Gemini model: {e}")
    
    async def generate_caption(self, request: CaptionGenerationRequest) -> CaptionGenerationResponse:
        """
        Generate AI-powered caption for social media content.
        """
        if not self.model:
            # Fallback response if AI is not available
            return self._fallback_caption_response(request)
        
        try:
            # Build the prompt
            prompt = self._build_caption_prompt(request)
            
            # Generate content
            response = self.model.generate_content(prompt)
            
            # Parse response
            caption_data = self._parse_gemini_response(response.text, request)
            
            return CaptionGenerationResponse(
                caption=caption_data["caption"],
                hashtags=caption_data["hashtags"],
                word_count=len(caption_data["caption"].split()),
                character_count=len(caption_data["caption"]),
                estimated_engagement=caption_data["engagement"]
            )
            
        except Exception as e:
            logger.error(f"Failed to generate caption: {e}")
            return self._fallback_caption_response(request)
    
    def _build_caption_prompt(self, request: CaptionGenerationRequest) -> str:
        """Build the prompt for Gemini API."""
        
        platform_context = ""
        if request.platforms:
            platform_names = [p.value.replace("_", " ").title() for p in request.platforms]
            platform_context = f" for {', '.join(platform_names)}"
        
        tone_descriptions = {
            ContentTone.CASUAL: "casual, friendly, and conversational",
            ContentTone.PROFESSIONAL: "professional, authoritative, and polished",
            ContentTone.FUN: "fun, playful, and entertaining",
            ContentTone.MOTIVATIONAL: "motivational, inspiring, and uplifting",
            ContentTone.EDUCATIONAL: "educational, informative, and helpful",
            ContentTone.TRENDY: "trendy, modern, and social media savvy"
        }
        
        niche_context = {
            ContentNiche.FITNESS: "fitness, health, and wellness",
            ContentNiche.FOOD: "food, cooking, and culinary experiences",
            ContentNiche.EDUCATION: "education, learning, and personal development",
            ContentNiche.LIFESTYLE: "lifestyle, daily life, and personal experiences",
            ContentNiche.BUSINESS: "business, entrepreneurship, and professional growth",
            ContentNiche.TECH: "technology, innovation, and digital trends"
        }
        
        hashtag_instruction = ""
        if request.include_hashtags:
            hashtag_instruction = f"""
- Include 5-10 relevant hashtags
- Mix popular and niche-specific hashtags
- Use hashtags relevant to {niche_context.get(request.niche, 'the content')}
"""
        
        prompt = f"""
Create an engaging social media caption{platform_context} based on this content:

Content Description: {request.video_description}
Niche: {niche_context.get(request.niche, request.niche.value)}
Tone: {tone_descriptions.get(request.tone, request.tone.value)}
Max Length: {request.max_length or 2200} characters

Requirements:
- Write in a {tone_descriptions.get(request.tone, request.tone.value)} tone
- Make it engaging and likely to get high engagement
- Keep it under {request.max_length or 2200} characters
- Focus on {niche_context.get(request.niche, 'the content')} audience
{hashtag_instruction}

Response Format (JSON):
{{
    "caption": "The main caption text",
    "hashtags": ["hashtag1", "hashtag2", "hashtag3"],
    "engagement": "high|medium|low"
}}

Generate the caption now:
"""
        
        return prompt
    
    def _parse_gemini_response(self, response_text: str, request: CaptionGenerationRequest) -> Dict[str, Any]:
        """Parse Gemini API response."""
        try:
            # Try to extract JSON from the response
            json_match = re.search(r'\{.*\}', response_text, re.DOTALL)
            if json_match:
                json_str = json_match.group()
                data = json.loads(json_str)
                
                # Validate required fields
                if "caption" in data and "hashtags" in data:
                    return {
                        "caption": data["caption"][:request.max_length or 2200],
                        "hashtags": data.get("hashtags", [])[:10],  # Limit to 10 hashtags
                        "engagement": data.get("engagement", "medium")
                    }
            
            # Fallback: treat entire response as caption
            lines = response_text.strip().split('\n')
            caption_lines = []
            hashtags = []
            
            for line in lines:
                line = line.strip()
                if line.startswith('#'):
                    # Extract hashtags
                    hashtags.extend(re.findall(r'#\w+', line))
                elif line and not line.startswith('{') and not line.startswith('}'):
                    caption_lines.append(line)
            
            caption = '\n'.join(caption_lines).strip()
            if not caption:
                caption = response_text.strip()
            
            return {
                "caption": caption[:request.max_length or 2200],
                "hashtags": list(set(hashtags))[:10],
                "engagement": "medium"
            }
            
        except Exception as e:
            logger.error(f"Failed to parse Gemini response: {e}")
            return self._get_fallback_data(request)
    
    def _fallback_caption_response(self, request: CaptionGenerationRequest) -> CaptionGenerationResponse:
        """Generate fallback response when AI is not available."""
        fallback_data = self._get_fallback_data(request)
        
        return CaptionGenerationResponse(
            caption=fallback_data["caption"],
            hashtags=fallback_data["hashtags"],
            word_count=len(fallback_data["caption"].split()),
            character_count=len(fallback_data["caption"]),
            estimated_engagement="medium"
        )
    
    def _get_fallback_data(self, request: CaptionGenerationRequest) -> Dict[str, Any]:
        """Get fallback caption and hashtags."""
        
        tone_templates = {
            ContentTone.CASUAL: "Just sharing this amazing moment! {description} What do you think?",
            ContentTone.PROFESSIONAL: "Excited to share: {description}. Looking forward to your thoughts and feedback.",
            ContentTone.FUN: "This is SO cool! {description} Who else loves this? 🔥",
            ContentTone.MOTIVATIONAL: "Remember: {description}. You've got this! Keep pushing forward! 💪",
            ContentTone.EDUCATIONAL: "Here's something interesting: {description}. Hope this helps you learn something new!",
            ContentTone.TRENDY: "Okay but like... {description} This is everything! ✨"
        }
        
        niche_hashtags = {
            ContentNiche.FITNESS: ["#fitness", "#workout", "#health", "#motivation", "#fitlife"],
            ContentNiche.FOOD: ["#food", "#foodie", "#cooking", "#recipe", "#delicious"],
            ContentNiche.EDUCATION: ["#education", "#learning", "#knowledge", "#skills", "#growth"],
            ContentNiche.LIFESTYLE: ["#lifestyle", "#life", "#daily", "#inspiration", "#vibes"],
            ContentNiche.BUSINESS: ["#business", "#entrepreneur", "#success", "#hustle", "#growth"],
            ContentNiche.TECH: ["#tech", "#technology", "#innovation", "#digital", "#future"]
        }
        
        caption_template = tone_templates.get(request.tone, tone_templates[ContentTone.CASUAL])
        caption = caption_template.format(description=request.video_description)
        
        base_hashtags = niche_hashtags.get(request.niche, ["#content", "#create", "#share"])
        common_hashtags = ["#viral", "#trending", "#follow", "#like", "#share"]
        
        hashtags = base_hashtags + common_hashtags[:3]
        
        return {
            "caption": caption,
            "hashtags": hashtags,
            "engagement": "medium"
        }


# Global AI service instance
ai_service = AIService() 